### PR TITLE
feat: add magic spline tool (#31)

### LINKED ISSUE #31: Magic wand splines
**Is your feature request related to a problem? Please describe.**

Being able to drag the mouse around an object and for a high-node-number spline to be created that autosnaps to the object boundaries.

**Describe the solution you'd like**

Button to toggle the feature on/off.

Using MouseDown + MouseMove to activate and drag the spline, use the local image gradient (i.e. within 10 pixels) and auto-drag the node to the peak in that gradient.

**Describe alternatives you've considered**

Bezier curves are an alternative for complex shapes but this is better because it's more reproducible and easier to use.

**Additional context**

### PR DESCRIPTION
**Pull request by ChasNelson1990**
# Description

Adds a magic spline tool. The user clicks the button and then presses down near the object of interest, they then drag the mouse (button still down) around the object and the spline will auto-snap to edges.

closes #31

# Dependency changes

None

# Testing

None added

# Documentation

None

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/toolboxes/spline/sobel.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      y: number,
+      r?: number,
+      g?: number,
+      b?: number,
+      alpha?: number
+    ): void {
+      r = r || 0;
+      g = g || 0;
+      b = b || 0;
+      alpha = alpha || 0;
```
Comment (cooper667) [inline src/toolboxes/spline/sobel.ts:39]:
```suggestion
    return function (
      x: number,
      y: number,
      r = 0,
      g = 0,
      b = 0,
      alpha = 0
    ): void {
```

I think this is the same and you type less?

File: src/toolboxes/spline/sobel.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+      setSobelPixelAt(x, y, magnitude, magnitude, magnitude, 255);
+    }
+  }
+
+  return {
+    width: width,
+    height: height,
+    data: sobelData,
+  };
```
Comment (cooper667) [inline src/toolboxes/spline/sobel.ts:100]:
```suggestion
  return {
    width,
    height,
    data: sobelData,
  };
```

File: src/toolboxes/spline/Canvas.tsx [hunk 3]
```diff
@@ -314,12 +312,8 @@ export class SplineCanvas extends Component<Props> {
...
       x,
       y,
-      this.props.imageWidth,
-      this.props.imageHeight,
-      this.props.scaleAndPan,
-      this.props.canvasPositionAndSize
-    );
-    const annotationData = this.props.annotationsObject.getActiveAnnotation();
+        this.props.imageData.width,
+        this.props.imageData.height,
```
Comment (philipjackson) [inline src/toolboxes/spline/Canvas.tsx:316]: I think there's an issue here (looking in the source for this commit, it's not just the diff viewer being weird: [https://github.com/gliff-ai/annotate/blob/026ad0fc6f6d8b19f0dba1970dc9a84dda66ee19/src/toolboxes/spline/Canvas.tsx](https://github.com/gliff-ai/annotate/blob/026ad0fc6f6d8b19f0dba1970dc9a84dda66ee19/src/toolboxes/spline/Canvas.tsx))

Comment (ChasNelson1990) [reply src/toolboxes/spline/Canvas.tsx:316]: Sorry, what is the issue?

Comment (ChasNelson1990) [reply src/toolboxes/spline/Canvas.tsx:316]: Oh, I think I see... I would ignore that... it doesn't exist in my local copy, must have accidental staged only half a segment

**Review by philipjackson — DISMISSED** [review]

**Review by SilviaZeta — DISMISSED** [review]

**Pull request closed without merging**